pkg/apis/core/validation/validation.go

```diff
@@ -4568,12 +4621,52 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 		return allErrs
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		//TODO(vinaykul,InPlacePodVerticalScaling): Investigate if PodStatus.QOSClass in kubelet, elsewhere.
+		//    Once we can rely on Status persistence, drop this block. QOSClass does not change
+		//    once it is bootstrapped on podCreate. This needs to be addressed before beta.
+		//    Ref: https://github.com/kubernetes/kubernetes/pull/102884/#discussion_r663280487
+		// reject attempts to change pod qos
+		oldQoS := qos.GetPodQOS(oldPod)
+		newQoS := qos.GetPodQOS(newPod)
+		if newQoS != oldQoS {
+			allErrs = append(allErrs, field.Invalid(fldPath, newQoS, "Pod QoS is immutable"))
+		}
+	}
+
 	// handle updateable fields by munging those fields prior to deep equal comparison.
 	mungedPodSpec := *newPod.Spec.DeepCopy()
 	// munge spec.containers[*].image
 	var newContainers []core.Container
 	for ix, container := range mungedPodSpec.Containers {
 		container.Image = oldPod.Spec.Containers[ix].Image // +k8s:verify-mutation:reason=clone
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
```

Comment:
We definitely don't want to allow further updates (unless the user rolls back to alpha and explicitly turns on this feature gate). The net effect of rolling back from beta to alpha is akin to turning feature-gate off (well not exactly, I expect code to significantly change between the two versions). This is an easy test to add in E2E. I'm tracking this for negative test cases we want to add before beta. Issue https://github.com/kubernetes/kubernetes/issues/114888

The corner-cases such as a pod resize was requested, kubelet "Deferred" the request, and then user rolls back kubelet and apiserver needs more thought. I will experiment with it to see what actually happens..

Reply:
I have added details on handling version skew. Please take a look.
https://github.com/kubernetes/enhancements/pull/3670

Reply:
I think I followed the table.  I was just looking for a sentence like "If the kubelet sees a Proposed change while its own gate is disabled, it will reject the change as Infeasible".  Version doesn't really matter - just gate state.

Reply:
I agree. And it will be implemented as such (in a follow-on PR to avoid resetting the review). In the table, I attempt to document expected behavior with allowed permutations of component versions (assuming that users haven't modified the default feature-gate value). In this case, 'A' in the table refers to [expected behavior](https://github.com/kubernetes/enhancements/pull/3670/files#diff-14542847beb0f0fd767db1aff1316f8569a968385e2bb89567c4cc0af1ae5942R931-R932) with those version combinations of kubelet and apiserver. The table just helps me keep things straight in my head :) 

Let me add a couple of lines before the table to better clarify the implementation regardless of versions.

Reply:
I hope the latest commit in the PR takes care of it. https://github.com/kubernetes/enhancements/pull/3670#event-8248704234